nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/reporting/AtlasNiFiFlowLineage.java

```diff
@@ -0,0 +1,714 @@
+                try (InputStream in = new FileInputStream(atlasPropertiesFile)) {
+                    atlasProperties.load(in);
+                }
+            } else {
+                final String fileInClasspath = "/" + ATLAS_PROPERTIES_FILENAME;
+                try (InputStream in = AtlasNiFiFlowLineage.class.getResourceAsStream(fileInClasspath)) {
+                    getLogger().info("Loading {} from classpath", new Object[]{fileInClasspath});
+                    if (in == null) {
+                        throw new ProcessException(String.format("Could not find %s in classpath." +
+                                " Please add it to classpath," +
+                                " or specify %s a directory containing Atlas properties file," +
+                                " or enable %s to generate it.",
+                                fileInClasspath, ATLAS_CONF_DIR.getDisplayName(), ATLAS_CONF_CREATE.getDisplayName()));
+                    }
+                    atlasProperties.load(in);
+                }
+            }
+        }
+
+        // Resolve default cluster name.
+        defaultClusterName = context.getProperty(ATLAS_DEFAULT_CLUSTER_NAME).evaluateAttributeExpressions().getValue();
+        if (defaultClusterName == null || defaultClusterName.isEmpty()) {
+            // If default cluster name is not specified by processor configuration, then load it from Atlas config.
+            defaultClusterName = atlasProperties.getProperty(ATLAS_PROPERTY_CLUSTER_NAME);
+        }
+
+        // If default cluster name is still not defined, processor should not be able to start.
+        if (defaultClusterName == null || defaultClusterName.isEmpty()) {
+            throw new ProcessException("Default cluster name is not defined.");
+        }
+
+        final AtlasAuthN atlasAuthN = getAtlasAuthN(atlasAuthNMethod);
+        atlasAuthN.configure(context);
+
+        // Create Atlas configuration file if necessary.
+        if (createAtlasConf) {
+
+            atlasProperties.put(ATLAS_PROPERTY_CLUSTER_NAME, defaultClusterName);
+            atlasProperties.put(ATLAS_PROPERTY_ENABLE_TLS, String.valueOf(isAtlasApiSecure));
+
+            setKafkaConfig(atlasProperties, context);
+
+            atlasAuthN.populateProperties(atlasProperties);
+
+            try (FileOutputStream fos = new FileOutputStream(atlasPropertiesFile)) {
+                String ts = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSX")
+                        .withZone(ZoneOffset.UTC)
+                        .format(Instant.now());
+                atlasProperties.store(fos, "Generated by Apache NiFi AtlasNiFiFlowLineage ReportingTask at " + ts);
+            }
+        }
+
+
+        atlasClient = NiFiAtlasClient.getInstance();
+        try {
+            atlasClient.initialize(urls.toArray(new String[]{}), atlasAuthN, confDir);
+        } catch (final NullPointerException e) {
```

Comment:
This seems odd to me, to catch a NullPointer - why would that be thrown? Should this be a more general Exception?

Reply:
Yeah, this is odd to me, too. However, AtlasClient throws NullPointerException when it can not find atlas properties file with following stack trace. Probably we should report this to Atlas project, too. Thanks.

```
2017-12-15 23:46:36,607 ERROR [StandardProcessScheduler Thread-4] o.a.n.c.s.StandardProcessScheduler Failed to invoke the On-Scheduled Lifecycle methods of [AtlasNiFiFlowLineage[id=5a7835aa-0160-1000-974d-14d95e0c38e9], java.lang.reflect.InvocationTargetException, 30 sec] due to {}; administratively yielding this ReportingTask and will attempt to schedule it again after {}
java.lang.reflect.InvocationTargetException: null
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:137)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:125)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotations(ReflectionUtils.java:70)
        at org.apache.nifi.util.ReflectionUtils.invokeMethodsWithAnnotation(ReflectionUtils.java:47)
        at org.apache.nifi.controller.scheduling.StandardProcessScheduler$2.run(StandardProcessScheduler.java:210)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.nifi.processor.exception.ProcessException: Failed to initialize Atlas client due to java.lang.NullPointerException. Make sure 'atlas-application.properties' is in the directory specified with Atlas Configuration Directory or under root classpath if not specified.
        at org.apache.nifi.atlas.reporting.AtlasNiFiFlowLineage.initAtlasClient(AtlasNiFiFlowLineage.java:505)
        at org.apache.nifi.atlas.reporting.AtlasNiFiFlowLineage.setup(AtlasNiFiFlowLineage.java:402)
        ... 16 common frames omitted
Caused by: java.lang.NullPointerException: null
        at org.apache.atlas.AtlasBaseClient.getClient(AtlasBaseClient.java:171)
        at org.apache.atlas.AtlasBaseClient.initializeState(AtlasBaseClient.java:154)
        at org.apache.atlas.AtlasBaseClient.initializeState(AtlasBaseClient.java:149)
        at org.apache.atlas.AtlasBaseClient.<init>(AtlasBaseClient.java:96)
        at org.apache.atlas.AtlasClientV2.<init>(AtlasClientV2.java:115)
        at org.apache.nifi.atlas.security.Basic.createClient(Basic.java:62)
        at org.apache.nifi.atlas.NiFiAtlasClient.initialize(NiFiAtlasClient.java:119)
        at org.apache.nifi.atlas.reporting.AtlasNiFiFlowLineage.initAtlasClient(AtlasNiFiFlowLineage.java:503)
        ... 17 common frames omitted
```

Reply:
Since the reason of NullPointer is known (it's possible to get due to different causes, but..) for NiFi users sake, I converted the Exception to more understandable error message.